### ISSUE TITLE
Too slow on 'big' tables even with cacheByIndexPath

### ISSUE THREAD

**Issue by oleynikd** [forkingdog/UITableView-FDTemplateLayoutCell#13]
In my app I usually have tables with ~150 rows.
Rows are very simple:
![screen shot 2015-04-30 at 3 11 47 am](https://cloud.githubusercontent.com/assets/3976868/7404088/d1bae69e-eee6-11e4-8b6a-3dab31b6affa.png)

When table view creates I have delay for about 3 seconds before it appears (iPhone 5 iOS 8.3)...
I tried to use new `- (CGFloat)fd_heightForCellWithIdentifier:(NSString *)identifier cacheByIndexPath:(NSIndexPath *)indexPath configuration:(void (^)(id cell))configuration;`  But seems like it starts caching heights only when I scroll through them for second time.

Also I noticed that `- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath` is called 3 times on init stage and all these 3 times heights are calculated from scratch... Here's my implementation:
```
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSLog(@"step 1: %@", indexPath);
    return [tableView fd_heightForCellWithIdentifier:@"DetailViewTableViewCell" cacheByIndexPath:indexPath configuration:^(iptvDetailedEPGTableViewCell *cell) {
        // Configure this cell with data, same as what you've done in "-tableView:cellForRowAtIndexPath:"
        NSLog(@"step 2: %@", indexPath); // Should not be called if using cached height, right?
        [self configureCell:cell forIndexPath:indexPath];
    }];
}
```

I get both log messages (3 * table length) times at init and than 1 more time when scrolling through the tableView. Only when scrolling second time I see only 1 log message.

Is it normal behaviour?

Thanks.

**Comment by sunnyxx** [comment]
This is UITableView's "bug like" problem in **iOS8** (not in iOS7), `-tableView:heightForRowAtIndexPath:` method will be called many times initially, and re-calculate when cell dequeues even you've scrolled through for more than one time, so this is why I provide a `cacheByIndexPath` API to avoid a mess of height calculations.  
When I implement `cacheByIndexPath`, I found it's not easy to automatically invalidate caches when data source updates. So I used a tricky way, swizzled a private method who triggers reload ( `-reloadData`, `-insertSection..`,`-deleteRows...` ... all work) and clear all current caches, it's a bit more complex, but makes this `cacheByIndexPath` API quite easy to use, no need for manual invalidation calls. But it has the **"3 times"** problem as you pointed, because UITableView itself triggers reload for 2 more times at init when it's being moved to window and layout its subviews. It's hard to solve unless I put the `invalidateCaches` method public.  

 So I'm thinking a better solution.
